src/blocks/agc.rs

```diff
@@ -0,0 +1,257 @@
+use num_complex::ComplexFloat;
+use futuresdr_pmt::Pmt;
+use futures::FutureExt;
+
+use crate::anyhow::Result;
+use crate::runtime::Block;
+use crate::runtime::BlockMeta;
+use crate::runtime::BlockMetaBuilder;
+use crate::runtime::Kernel;
+use crate::runtime::MessageIo;
+use crate::runtime::MessageIoBuilder;
+use crate::runtime::StreamIo;
+use crate::runtime::StreamIoBuilder;
+use crate::runtime::WorkIo;
+
+
+pub struct AGC<T> {
+    // Minimum value that has to be reached in order for AGC to start adjusting gain.
+    squelch: T,
+    // maximum gain value (0 for unlimited).
+    max_gain: T,
+    // initial gain value.
+    gain: T,
+    // reference value to adjust signal power to.
+    reference_power: T,
+    // the update rate of the loop.
+    update_rate: T,
+    // (Boolean) Set when gain should not be adjusted anymore, but rather be locked to the current value
+    gain_lock: u32,
+}
+
+impl<T> AGC<T>
+    where
+        T: Send + Sync + ComplexFloat + 'static,
+{
+    pub fn new(
+        squelch: T,
+        max_gain: T,
+        gain: T,
+        update_rate: T,
+        reference_power: T,
+        gain_lock: u32,
+    ) -> Block {
+        Block::new(
+            BlockMetaBuilder::new("AGC").build(),
+            StreamIoBuilder::new()
+                .add_input::<T>("in")
+                .add_output::<T>("out")
+                .build(),
+            MessageIoBuilder::<Self>::new()
+                .add_input("gain_lock",
+                           |block: &mut AGC<T>,
+                            _mio: &mut MessageIo<AGC<T>>,
+                            _meta: &mut BlockMeta,
+                            p: Pmt| {
+                               async move {
+                                   if let Pmt::U32(ref r) = &p {
+                                       block.gain_lock = *r;
+                                   }
+                                   Ok(p)
+                               }.boxed()
+                           },
+                )
+                .add_input("max_gain",
+                           |block: &mut AGC<T>,
+                            _mio: &mut MessageIo<AGC<T>>,
+                            _meta: &mut BlockMeta,
+                            p: Pmt| {
+                               async move {
+                                   if let Pmt::Any(ref r) = &p {
```

Comment:
If member types are just f32 then this would be much easier with just Pmt::F32(), no Pmt::Any complications.